cirq/qis/measures.py

```diff
@@ -138,9 +138,17 @@ def fidelity(
         )
     state1 = quantum_state(state1, qid_shape=qid_shape, validate=validate, atol=atol)
     state2 = quantum_state(state2, qid_shape=qid_shape, validate=validate, atol=atol)
-    state1 = state1.density_matrix() if state1._is_density_matrix() else state1.state_vector()
-    state2 = state2.density_matrix() if state2._is_density_matrix() else state2.state_vector()
-    return _fidelity_state_vectors_or_density_matrices(state1, state2)
+    state1_arr = (
+        state1.density_matrix()
+        if state1._is_density_matrix()
+        else cast(np.ndarray, state1.state_vector())
+    )
+    state2_arr = (
+        state2.density_matrix()
+        if state2._is_density_matrix()
+        else cast(np.ndarray, state2.state_vector())
```

Comment:
Optional: IIUC this is just to convert from `Optional[np.ndarray]` to `np.ndarray`. It'd be nice if you could make it more apparent to the reader. Perhaps an `assert` is understood by both mypy and humans?

Reply:
assert would work but not in this ternary operation

I found this thread when going through these: https://github.com/python/typing/issues/645

honestly, I think these code patterns indicate that there's a missing method. Either "`state_vector_for_sure_because_i_already_checked`" or `array_form_irrespective_of_whether_its_density_or_state`. I'll investigate

Reply:
Otherwise the cast is safe since the only time `state_vector` returns none is when `_is_density_matrix` is true. Obviously mypy isn't smart enough to deduce that full logic

Reply:
I agree with you that there is a missing function to aid this pattern. Perhaps something like Rust's [unwrap](https://doc.rust-lang.org/rust-by-example/error/option_unwrap.html)? This would have the benefit of being useful for all `Optional` types, not just state vectors. Ideally, we'd use it similar to how you're using `cast` here:

```python
state2_arr = (
        state2.density_matrix()
        if state2._is_density_matrix()
        else unwrap(state2.state_vector())
```

WDYT?

Reply:
I like it.

nit: maybe `state_or_density_array` -> `state_vector_or_density_matrix`. The latter is clearer (especially since `state_tensor` is also a thing so `state` alone in this context is ambiguous) and more consistent with other method names.

Reply:
sure. I was trying to think of names that weren't *too* long but I guess this is a somewhat niche function so it can handle a long name.